File: Data Structures & Algorithms/anagram-groups/submission-1.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, List<String>> idxHMap = new HashMap<>();
        for(String s : strs) {
            String sCharCount = countCharacters(s);
            if(idxHMap.containsKey(sCharCount)) {
                idxHMap.get(sCharCount).add(s);
            }
            else {
                List<String> sList = new ArrayList<>();
                sList.add(s);
                idxHMap.put(sCharCount, sList);
            }
        }

        // Loop over enery element in the hashmap
        return new ArrayList<>(idxHMap.values());
    }

    private String countCharacters(String s) {
        int[] countChar = new int[26];

        int length = s.length();
        for(int i = 0  ; i < length ; i++) {
            char ch = s.charAt(i);
            int idx = ch - 'a';
            countChar[idx]++;
        }

        StringBuilder sb = new StringBuilder(0);
        for(int i = 0 ; i < countChar.length ; i++) {
            sb.append((char)(i +  'a'));
            sb.append(countChar[i]);
        }

        return sb.toString();
    }
}

File: Data Structures & Algorithms/anagram-groups/submission-1.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        char_freq = {}
        for s in strs:
            fr = self.freq(s)
            if fr in char_freq:
               char_freq[fr].append(s)
            else:
                char_freq[fr] = [s]
        
        return list(char_freq.values())

    def freq(self, s: str) -> dict:
        freq = {}

        for x in s:
            if x in freq:
                freq[x] = freq[x] + 1
            else:
                freq[x] = 1

        return tuple(sorted(freq.items()))